feat: the unusedSetOptionIn linter

adomani opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

adomani committed:
feat: the `unusedSetOptionIn` linter

adomani committed:
lints

grunweg commented on Mathlib/Tactic/Linter/UnusedSetOptionIn.lean:
```diff
@@ -0,0 +1,65 @@
+/-- converts
+* `theorem x ...` to  `some (example ... , x)`,
+* `lemma x ...`   to  `some (example ... , x)`,
+* `example ...`   to ``some (example ... , `example)``,
```
Is this correct?
```suggestion
* `example ...`   to ``some (example ... , 'example')`,

adomani replied:
I ended up going for your version: the reason behind what I had, was that I was intepreting the `example` as a `Name`, but it does not matter much either way -- it is a syntax node anyway!

grunweg commented on Mathlib/Tactic/Linter/UnusedSetOptionIn.lean:
```diff
@@ -0,0 +1,65 @@
+import Lean.Linter.Util
+
+/-!
+The `unusedSetOptionIn` linter warns against "most external" unused `set_option ... in`s.
```
how about "... against outer-most `set option ... in`s which are not necessary"?

adomani replied:
Also overhauled by the new docs on the linter.

grunweg commented on Mathlib/Tactic/Linter/UnusedSetOptionIn.lean:
```diff
@@ -0,0 +1,65 @@
+    return some (← `($dm:declModifiers example $ds $dv:declVal), mkIdent `example)
+  | _ => return none
+
+/-- The `unusedSetOptionIn` linter warns against "most external" unused `set_option ... in`s. -/
```
... possibly with the same phrasing

adomani replied:
Yes, I should have recycled this using `inherit_doc`.

grunweg commented on Mathlib/Tactic/Linter/UnusedSetOptionIn.lean:
```diff
@@ -0,0 +1,65 @@
+/-- Gets the value of the `linter.unusedSetOptionIn` option. -/
+def getSetOptionIn (o : Options) : Bool := Linter.getLinterValue linter.unusedSetOptionIn o
+
+/-- reports a warning if the "first layer" `set_option ... in` is unused. -/
```
How about this tweak?
"Report a warning if a `set_option ... in` command is unnecessary (i.e., the code elaborates fine without).
We only report the outermost `set_option ... in` (i.e., superfluous `set_option ... in`s which are nested are not linted against)."

If you'd like, you could also add
- why is not reporting on nested superfluous set_option in is not not so bad (is this just for simplicity, as the linter can be run multiple times? that's probably fine!)
- add one sentence why this is bad: for instance, might distract or confuse --- if this option was for backwards incompatibility, the fact that it can be removed now is a useful signal.

adomani replied:
The new doc-string is (once I will push)
```
Report a warning if a `set_option ... in` command is unnecessary
(i.e., the code elaborates fine without it).
We only report the outermost `set_option ... in` (i.e., nested, superfluous `set_option ... in`s
are not linted against).

This is useful since unnecessary `set_option ... in`s are often silent left-overs of adaptations
that are no longer needed.
Cleaning them up, helps maintaining healthy, up-to-date code.

The linter runs just on the outermost `set_option ... in` mostly for simplicity.
Consider whether emitting a warning if one of the nested `set_option ... in`s can be omitted,
by removing them one at a time.
```

grunweg replied:
Looks great to me!

grunweg commented on Mathlib/Tactic/Linter/UnusedSetOptionIn.lean:
```diff
@@ -0,0 +1,65 @@
+  let s ← get
+  match stx with
+    | .node _ ``Lean.Parser.Command.in #[
+        .node _ ``Lean.Parser.Command.set_option _,
```
Do you want to differentiate which option is set? For instance, if I set `profiler.threshold 50 in`, I want to have this for temporary purposes - and linting on it seems more dubious than useful.
(The `set_option` linter also covers that case - written in Python, but its rewrite in Lean is awaiting review.)

adomani replied:
I was planning to discriminate on which option is being set by going over what the lniter flags.  For instance, I am not sure whether this linter should handle various thresholds/heartBeats, since maybe a more dedicated one could also do some diagnostics on the side.

grunweg reviewed (commented):
Great idea! I know this is in draft status - I took a look anyway (and hope you don't mind).
I have a few suggestions for the docs (see below). I also wonder if "unnecessarySetOptionIn" would be a better name - since you're not checking if the option *never influences* elaboration, but merely if *things work without*.

grunweg commented:
The linter error in `Mathlib/Order/LiminfLimsup.lean:1331` is a false positive: with removing the set_option, compilation succeeds, but with a *warning*. (By the way, this could be another nice test case: checking e.g. that in a lemma using a deprecated declaration, the `set_option linter.deprecated false in` is not considered superfluous.)

grunweg commented:
Another interesting puzzle: the linter does not lint the `set_option genSizeOfSpec false in` in `SetTheory/Ordinal/Notation.lean`.
Removing that and running #lint seems to work... but perhaps that's a different class of issue than what this linter is for!

adomani committed:
report the option

grunweg committed:
cherry-pick from 13655: some fixes

grunweg committed:
A few more tests.

adomani committed:
docs

grunweg committed:
Fix a few more errors: close to all the linter finds, right now.
- 9 mentions of 12532 remain
- I did not investigate the heartbeat removals at all

grunweg committed:
Revert change to AffineScheme: the bump is needed, somehow.

adomani committed:
lint also defs and instances, do not lint linters + tests

adomani committed:
Merge branch 'adomani/unused_set_option_in_linter' of github.com:leanprover-community/mathlib4 into adomani/unused_set_option_in_linter

adomani committed:
docs

grunweg committed:
Some more.

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/unused_set_option_in_linter

adomani committed:
Merge branch 'master' into adomani/unused_set_option_in_linter

adomani committed:
Merge branch 'master' into adomani/unused_set_option_in_linter

adomani committed:
imports

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/unused_set_option_in_linter

adomani committed:
imports

adomani committed:
rename file Unused --> Unnecessary, use syntax matching

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/unused_set_option_in_linter

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/unused_set_option_in_linter

adomani committed:
chore: remove unused set_options

adomani committed:
Merge branch 'adomani/remove_set_options' into adomani/unused_set_option_in_linter

adomani committed:
remove more

adomani committed:
Merge branch 'adomani/remove_set_options' into adomani/unused_set_option_in_linter

adomani committed:
silence the linter

adomani committed:
do not lint MathlibTest

adomani committed:
do not lint MathlibTest

adomani committed:
lint the test

adomani committed:
really lint

adomani committed:
revert one heartbeat change since it counts 185348 heartbeats

adomani committed:
Merge branch 'adomani/remove_set_options' into adomani/unused_set_option_in_linter

adomani committed:
ignore heartbeats by default

adomani committed:
add docs

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/unused_set_option_in_linter

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.
